store/src/event.rs

```diff
@@ -0,0 +1,142 @@
+    let get_block = |cid: &Cid| -> Result<&Vec<u8>> {
+        if let Some(bytes) = blocks.get(cid) {
+            Ok(bytes)
+        } else {
+            Err(Error::new_fatal(            anyhow!(
+                "event car imcomplete, all referenced blocks should be availble in the car data. missing cid: {cid}"
+            )))
+        }
+    };
+
+    let event_bytes = get_block(&event_cid)?;
+    let event: unvalidated::Event = serde_ipld_dagcbor::from_slice(&event_bytes)
+        .context("decoding event")
+        .map_err(Error::new_fatal)?;
+    match event {
+        unvalidated::Event::Signed(event) => {
+            let link = event
+                .link()
+                .ok_or_else(|| anyhow!("event should have a link"))
+                .map_err(Error::new_fatal)?;
+
+            let payload_bytes = get_block(&link)?;
+            let payload: unvalidated::Payload = serde_ipld_dagcbor::from_slice(&payload_bytes)
+                .context("decoding payload")
+                .map_err(Error::new_fatal)?;
+            match payload {
+                unvalidated::Payload::Init(payload) => Ok(EventId::new(
+                    &network,
+                    SEP_KEY,
+                    payload.header().model().as_slice(),
+                    &payload
+                        .header()
+                        .controllers()
+                        .get(0)
+                        .ok_or_else(|| anyhow!("init payload should have at least one controller"))
+                        .map_err(Error::new_fatal)?,
+                    &event_cid,
+                    &event_cid,
+                )),
+                unvalidated::Payload::Data(payload) => {
+                    let header = payload
+                        .header()
+                        .ok_or_else(|| anyhow!("event payload should have a header"))
+                        .map_err(Error::new_fatal)?;
+                    Ok(EventId::new(
+                        &network,
+                        SEP_KEY,
+                        header.model().as_slice(),
+                        &header
+                            .controllers()
+                            .ok_or_else(|| {
+                                anyhow!("event payload should have a set of controllers")
+                            })
+                            .map_err(Error::new_fatal)?
+                            .get(0)
+                            .ok_or_else(|| {
+                                anyhow!("event payload should have at least one controller")
+                            })
+                            .map_err(Error::new_fatal)?,
+                        payload.id(),
+                        &event_cid,
+                    ))
+                }
+            }
+        }
+        unvalidated::Event::Time(event) => {
+            let init = event.id();
+            // TODO should time event CAR files contain their init event?
```

Comment:
wouldn't this same problem exist for data events too?  I think you just have to have received the init event independently before you can even really start parsing or validating data and time events.